machineconfiguration/v1/types.go

```diff
@@ -503,8 +503,75 @@ type MachineConfigPoolStatus struct {
 	// +listType=atomic
 	// +optional
 	CertExpirys []CertExpiry `json:"certExpirys"`
+
+	// poolSynchronizersStatus is the status of the machines managed by the pool synchronizers.
+	// +openshift:enable:FeatureGate=PinnedImages
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=poolSynchronizerType
+	// +optional
+	PoolSynchronizersStatus []PoolSynchronizerStatus `json:"poolSynchronizersStatus"`
+}
+
+// +kubebuilder:validation:XValidation:rule="self.machineCount >= self.updatedMachineCount", message="machineCount must be greater than or equal to updatedMachineCount"
+// +kubebuilder:validation:XValidation:rule="self.machineCount >= self.availableMachineCount", message="machineCount must be greater than or equal to availableMachineCount"
+// +kubebuilder:validation:XValidation:rule="self.machineCount >= self.unavailableMachineCount", message="machineCount must be greater than or equal to unavailableMachineCount"
+// +kubebuilder:validation:XValidation:rule="self.machineCount >= self.readyMachineCount", message="machineCount must be greater than or equal to readyMachineCount"
+// +kubebuilder:validation:XValidation:rule="self.availableMachineCount >= self.readyMachineCount", message="availableMachineCount must be greater than or equal to readyMachineCount"
+type PoolSynchronizerStatus struct {
+	// name is the name of a pool synchronizer. Must adhere to
+	// RFC-1123 (https://tools.ietf.org/html/rfc1123).
+	// Made up of one of more period-separated (.) segments, where each segment
+	// consists of alphanumeric characters and hyphens (-), must begin and end
+	// with an alphanumeric character, and is at most 63 characters in length.
+	// The total length of the name must not exceed 253 characters.
+	// +openshift:enable:FeatureGate=PinnedImages
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```

Comment:
this is an open question to the team do you want to support the option for multiple synchronizers per type? Currently, the only type is PinnedImageSets. We could instead populate a list of references that make up the synchronizer.

```
// MCOObjectReference holds information about an object the MCO either owns
// or modifies in some way
type MCOObjectReference struct {
	// name is the object name.
	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
	// It may consist of only alphanumeric characters, hyphens (-) and periods (.)
	// and must be at most 253 characters in length.
	// +kubebuilder:validation:MaxLength:=253
	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
	// +kubebuilder:validation:Required
	Name string `json:"name"`
}
```

Reply:
It makes sense to me to refer multiple PinnedImageSets by a synchronizer type.
If I understand correctly, this would mean that we would be able to express PinnedImageSet related information in MCP something like this:
```
apiVersion: machineconfiguration.openshift.io/v1
        kind: MachineConfigPool
        spec:
          pinnedImageSets:
            - name: test-pinnedimageset
            - name: test1-pinnedimageset
            - name: test2-pinnedimageset
        status:
          poolSynchronizersStatus:
            - source: 
              - name: test-pinnedimageset
              - name: test1-pinnedimageset
              - name: test2-pinnedimageset
              poolSynchronizerType: PinnedImageSets
              machineCount: 3
              updatedMachineCount: 3
              readyMachineCount: 3
              availableMachineCount: 3
              unavailableMachineCount: 0
              observedGeneration: 3
```

Reply:
thanks, I find this as a useful change since this provides PinnedImageSet related status more user friendly to read.

Reply:
The possible downside is a pool can declare up to 100 pinnedImageSets plus to be complete we would need to report on the current generation as well?

Reply:
I thought that's what we are anyway doing by allowing users to create and specify multiple pinnedImageSets targeting a pool. Are we limiting number of pinnedImageSets user can specify in spec? I may have missed if any changes were made.

Reply:
second proposal

```
apiVersion: machineconfiguration.openshift.io/v1
        kind: MachineConfigPool
        spec:
          pinnedImageSets:
            - name: test-pinnedimageset
            - name: test1-pinnedimageset
            - name: test2-pinnedimageset
        status:
          poolSynchronizersStatus:
            - source: 
              - name: test-pinnedimageset
                desiredGeneration: 3
                currentGeneration: 2
              - name: test1-pinnedimageset
                desiredGeneration: 3
                currentGeneration: 2
              - name: test2-pinnedimageset
                desiredGeneration: 3
                currentGeneration: 2
              poolSynchronizerType: PinnedImageSets
              machineCount: 3
              updatedMachineCount: 3
              readyMachineCount: 3
              availableMachineCount: 3
              unavailableMachineCount: 0
              observedGeneration: 3
  ```